### DISCUSSION TITLE
Unable to type asyncio.Lock as AbstractAsyncContextManager

### DISCUSSION THREAD

**Discussion by FasterSpeeding** [microsoft/pyright#2435] [General]
```py
import asyncio
import contextlib
import typing

def foo() -> contextlib.AbstractAsyncContextManager[typing.Any]:
    return asyncio.Lock()
```
While linting this code with pyright 1.1.178 I get the following issue error
```py
C:\Users\Lucina\PycharmProjects\tanjun\test.py
  C:\Users\Lucina\PycharmProjects\tanjun\test.py:6:12 - error: Expression of type "Lock" cannot be assigned to return type "AbstractAsyncContextManager"
    "Lock" is incompatible with protocol "AbstractAsyncContextManager"
      "__aexit__" is an incompatible type
        Type "(exc_type: Type[BaseException] | None, exc: BaseException | None, tb: TracebackType | None) -> Awaitable[None]" cannot be assigned to type "(exc_type: Type[BaseException] | None, exc_value: BaseException | None, traceback: TracebackType | None) -> Awaitable[bool | None]"
          Parameter name mismatch: "exc_value" versus "exc"
          Parameter name mismatch: "traceback" versus "tb" (reportGeneralTypeIssues)
1 error, 0 warnings, 0 infos
Completed in 0.574sec
```

While this is technically correct this doesn't match the runtime semantics of context managers (where all of these values are passed as positional arguments) which make Lock a valid async context manager but I do see that it isn't type safe to just mix and match names like that with arguments which can be passed both positionally or by their name.

For reference MyPy allows this on strict and python's stubs are accurate as AbstractAsyncContextManager doesn't enforce that those parameters are passed positionally.

To be honest I wasn't exactly sure where this belonged so I thought just a general discussion here would be good enough (or at least better than just not bringing it up at all).

**Comment by JelleZijlstra** [comment]
We can probably fix this in typeshed by making the args to AbstractAsyncContextManager positional-only. I submitted python/typeshed#6172 to do that.

**Comment by erictraut** [comment]
This is a legitimate type error. I'm not sure why mypy doesn't report it.

This is a bug in the type stub. Looks like @JelleZijlstra is on top of it already. Thanks!

**Reply by JelleZijlstra** [reply]
I believe mypy generally ignores argument names in protocols, which is technically unsafe but fine in practice most of the time.
